lib/kino/smart_cell/big_query.ex

```diff
@@ -0,0 +1,112 @@
+defmodule Kino.SmartCell.BigQuery do
+  @moduledoc false
+
+  # A smart cell used to establish connection to a database.
+
+  use Kino.JS, assets_path: "lib/assets/big_query"
+  use Kino.JS.Live
+  use Kino.SmartCell, name: "Google BigQuery database connection"
+
+  alias Kino.SmartCell.BigQuery.Connection
+
+  @impl true
+  def init(attrs, ctx) do
+    fields = %{
+      "variable" => Kino.SmartCell.prefixed_var_name("conn", attrs["variable"]),
+      "project_id" => attrs["project_id"] || "",
+      "dataset" => attrs["dataset"] || "",
+      "private_key_id" => attrs["private_key_id"] || "",
+      "private_key" => attrs["private_key"] || "",
+      "client_email" => attrs["client_email"] || "",
+      "client_id" => attrs["client_id"] || "",
+      "client_x509_cert_url" => attrs["client_x509_cert_url"] || ""
+    }
+
+    {:ok, assign(ctx, fields: fields, missing_dep: missing_dep())}
+  end
+
+  @impl true
+  def handle_connect(ctx) do
+    payload = %{
+      fields: ctx.assigns.fields,
+      missing_dep: ctx.assigns.missing_dep
+    }
+
+    {:ok, payload, ctx}
+  end
+
+  @impl true
+  def handle_event("update_field", %{"field" => field, "value" => value}, ctx) do
+    updated_fields = to_updates(ctx.assigns.fields, field, value)
+    ctx = update(ctx, :fields, &Map.merge(&1, updated_fields))
+
+    ctx =
+      if missing_dep() == ctx.assigns.missing_dep do
+        ctx
+      else
+        broadcast_event(ctx, "missing_dep", %{"dep" => missing_dep()})
+        assign(ctx, missing_dep: missing_dep())
+      end
+
+    broadcast_event(ctx, "update", %{"fields" => updated_fields})
+
+    {:noreply, ctx}
+  end
+
+  defp to_updates(fields, "variable", value) do
+    if Kino.Utils.Code.valid_variable_name?(value) do
+      %{"variable" => value}
+    else
+      %{"variable" => fields["variable"]}
+    end
+  end
+
+  defp to_updates(_fields, field, value), do: %{field => value}
+
+  @impl true
+  def to_attrs(ctx) do
+    ctx.assigns.fields
+  end
+
+  @impl true
+  def to_source(attrs) do
+    attrs
+    |> to_quoted()
+    |> Kino.Utils.Code.quoted_to_string()
+  end
+
+  defp to_quoted(attrs) do
+    quote do
+      alias unquote(Connection), as: BigQuery
```

Comment:
An important aspect of smart cells is teaching people how to write the code that they would need to write otherwise, so we should minimize hiding it behind `Kino.*` functions.

Looking at the code, directly we could generate this:

```elixir
scopes = ["https://www.googleapis.com/auth/cloud-platform"]

credentials = %{
  "type" => "...",
  "project_id" => "...",
  "private_key_id" => "...",
  "private_key" => "...",
  "client_email" => "...",
  "client_id" => "...",
  "auth_uri" => "...",
  "token_uri" => "...",
  "auth_provider_x509_cert_url" => "...",
  "client_x509_cert_url" => "..."
}

{:ok, goth} =
  Kino.start_child({Goth, source: {:service_account, credentials, [scopes: scopes]}})

{:ok, %{token: token}} = Goth.fetch(goth)
conn = GoogleApi.BigQuery.V2.Connection.new(token)
```

The only issue is that `conn` is a Finch client, so we can't really tell that it's the db connection. Also, it doesn't include some of the information necessary for making queries. So to integrate it into the SQL cell, we need a more informative struct, which I think gives two options:

* create a plain struct `%Kino.SmartCell.DbConnection.BigQuery{conn: conn}` and put any other relevant information there
* create a separate package specifically for Big Query that makes the HTTP requests directly

@josevalim any other ideas?

Reply:
Yeah, that's what I had in mind with the second option. It wouldn't be Livebook/Kino specific in any way, but would provide a more straightforward API, with a proper connection struct. Basically it should be something usable on its own that we only use to generate the code. It's better to avoid this kind of of micro-packages, but in this case it would be useful.

Reply:
> The only issue is that `conn` is a Finch client

This is a bigger issue, because if I get the `{:ok, conn} = Finch...`, conn will be a PID which doesn't work with Finch's library. They asks for the GenServer's name that we defined with `Kino.start_link/1`. So, even with Finch, I need to export a variable (`conn`) with the module name.

```elixir
scopes = ["https://www.googleapis.com/auth/cloud-platform"]

credentials = %{
  "type" => "...",
  "project_id" => "...",
  "private_key_id" => "...",
  "private_key" => "...",
  "client_email" => "...",
  "client_id" => "...",
  "auth_uri" => "...",
  "token_uri" => "...",
  "auth_provider_x509_cert_url" => "...",
  "client_x509_cert_url" => "..."
}

conn = BigQuery

{:ok, goth} = Goth.start_link(source: {:service_account, credentials, scopes: scopes})
{:ok, _finch_pid} = Kino.start_child({Finch, name: BigQuery})
```

And, to request:

```elixir
headers = []
body = nil

request = Finch.build(:get, "https://www.google.com.br", headers, body, [])
{:ok, _response} = Finch.request(request, conn)
```

A simple example, but in my guess, this is less user-friendly than before.

Reply:
> create a plain struct `%Kino.SmartCell.DbConnection.BigQuery{conn: conn}` and put any other relevant information there

Oh, actually it doesn't need to be a struct, just anything that we can clearly recognize as a big query connection. So it could be a regular map like `conn = %{big_query_conn: conn, ...}`. That we can easily match in the SQL cell and generate the code accordingly. That's slightly indirect, but a pretty reasonable option IMO.

Reply:
This is what needs to be used to perform the same request as the `google_api_big_query` lib

![image](https://user-images.githubusercontent.com/6402997/164250402-40735851-9324-45d9-ba92-09c68a9bfae8.png)

Reply:
One other piece to the puzzle is that we definitely need a struct for the response. With the current APIs we have `GoogleApi.BigQuery.V2.Model.QueryRequest`, which is good, otherwise we will need to come up with something. That's because we will soon have the [table](https://github.com/dashbitco/table) package, which means we can implement a protocol for the response struct, and there we will handle all the type conversion, etc. So we won't need the post-response steps in the query code :D